receiver/k8seventsreceiver/receiver.go

```diff
@@ -0,0 +1,152 @@
+	stopperChanList []chan struct{}
+	startTime       time.Time
+	ctx             context.Context
+	cancel          context.CancelFunc
+	obsrecv         *obsreport.Receiver
+}
+
+// newReceiver creates the Kubernetes events receiver with the given configuration.
+func newReceiver(
+	set component.ReceiverCreateSettings,
+	config *Config,
+	consumer consumer.Logs,
+	client k8s.Interface,
+) (component.LogsReceiver, error) {
+	transport := "http"
+
+	return &k8seventsReceiver{
+		settings:     set,
+		config:       config,
+		client:       client,
+		logsConsumer: consumer,
+		startTime:    time.Now(),
+		obsrecv: obsreport.NewReceiver(obsreport.ReceiverSettings{
+			ReceiverID:             config.ID(),
+			Transport:              transport,
+			ReceiverCreateSettings: set,
+		}),
+	}, nil
+}
+
+func (kr *k8seventsReceiver) Start(ctx context.Context, host component.Host) error {
+	kr.ctx, kr.cancel = context.WithCancel(ctx)
+
+	kr.settings.Logger.Info("starting to watch namespaces for the events.")
+	if len(kr.config.Namespaces) == 0 {
+		kr.startWatch(corev1.NamespaceAll)
+	} else {
+		for _, ns := range kr.config.Namespaces {
+			kr.startWatch(ns)
+		}
+	}
+
+	return nil
+}
+
+func (kr *k8seventsReceiver) Shutdown(context.Context) error {
+	// Stop watching all the namespaces by closing all the stopper channels.
+	for _, stopperChan := range kr.stopperChanList {
+		close(stopperChan)
+	}
+	kr.cancel()
+	return nil
+}
+
+// Add the 'New Event' handler and trigger the watch for a specific namespace.
+func (kr *k8seventsReceiver) startWatch(ns string) {
+	stopperChan := make(chan struct{})
+	kr.stopperChanList = append(kr.stopperChanList, stopperChan)
+	kr.startWatchingNamespace(kr.client, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			ev := obj.(*corev1.Event)
+			kr.handleNewEvent(ev)
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			ev := obj.(*corev1.Event)
+			kr.handleNewEvent(ev)
+		},
+	}, ns, stopperChan)
+}
+
+func (kr *k8seventsReceiver) handleNewEvent(ev *corev1.Event) {
+	if kr.allowEvent(ev) {
+		ld := k8sEventToLogData(kr.settings.Logger, ev)
+
+		ctx := kr.obsrecv.StartLogsOp(kr.ctx)
+		consumerErr := kr.logsConsumer.ConsumeLogs(ctx, ld)
+		kr.obsrecv.EndLogsOp(ctx, typeStr, 1, consumerErr)
+	}
+}
+
+// startWatchingNamespace creates an informer and starts
+// watching a specific namespace for the events.
+func (kr *k8seventsReceiver) startWatchingNamespace(
+	clientset k8s.Interface,
+	handlers cache.ResourceEventHandlerFuncs,
+	ns string,
+	stopper chan struct{},
+) {
+	client := clientset.CoreV1().RESTClient()
+	watchList := cache.NewListWatchFromClient(client, "events", ns, fields.Everything())
+	_, controller := cache.NewInformer(watchList, &corev1.Event{}, 0, handlers)
+	go controller.Run(stopper)
+}
+
+// Allow events with eventTimestamp(EventTime/LastTimestamp/FirstTimestamp)
+// not older than the receiver start time by 'InitialLookback' so that
+// event flood can be avoided upon startup.
```

Comment:
Please update the comment

Reply:
I have updated all the relevant comments accordingly.